core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/DefaultCompressionAlgorithm.java

```diff
@@ -0,0 +1,354 @@
+ * {@code createCompressionStream} and {@code createDecompressionStream} with non default buffer
+ * sizes, the configuration object must be changed. In this case, concurrent calls to
+ * {@code createCompressionStream} and {@code createDecompressionStream} would mutate the
+ * configuration object beneath each other, requiring synchronization to avoid undesirable activity
+ * via co-modification. To avoid synchronization entirely, we will create Codecs with their own
+ * Configuration object and cache them for re-use. A default codec will be statically created, as
+ * mentioned above to ensure we always have a codec available at loader initialization.
+ * <p>
+ * There is a Guava cache defined within Algorithm that allows us to cache Codecs for re-use. Since
+ * they will have their own configuration object and thus do not need to be mutable, there is no
+ * concern for using them concurrently; however, the Guava cache exists to ensure a maximal size of
+ * the cache and efficient and concurrent read/write access to the cache itself.
+ * <p>
+ * To provide Algorithm specific details and to describe what is in code:
+ * <p>
+ * LZO will always have the default LZO codec because the buffer size is never overridden within it.
+ * <p>
+ * LZ4 will always have the default LZ4 codec because the buffer size is never overridden within it.
+ * <p>
+ * GZ will use the default GZ codec for the compression stream, but can potentially use a different
+ * codec instance for the decompression stream if the requested buffer size does not match the
+ * default GZ buffer size of 32k.
+ * <p>
+ * Snappy will use the default Snappy codec with the default buffer size of 64k for the compression
+ * stream, but will use a cached codec if the buffer size differs from the default.
+ */
+public class DefaultCompressionAlgorithm extends Configured {
+
+  public static class FinishOnFlushCompressionStream extends FilterOutputStream {
+
+    FinishOnFlushCompressionStream(CompressionOutputStream cout) {
+      super(cout);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+      out.write(b, off, len);
+    }
+
+    @Override
+    public void flush() throws IOException {
+      CompressionOutputStream cout = (CompressionOutputStream) out;
+      cout.finish();
+      cout.flush();
+      cout.resetState();
+    }
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultCompressionAlgorithm.class);
+
+  /**
+   * Guava cache to have a limited factory pattern defined in the Algorithm enum.
+   */
+  private static LoadingCache<Entry<DefaultCompressionAlgorithm,Integer>,
+      CompressionCodec> codecCache =
+          CacheBuilder.newBuilder().maximumSize(25).build(new CacheLoader<>() {
+            @Override
+            public CompressionCodec load(Entry<DefaultCompressionAlgorithm,Integer> key) {
+              return key.getKey().createNewCodec(key.getValue());
+            }
+          });
```

Comment:
Do we want this cache here? It looks like before we had one cache in Compression class and now you expanded it out to have a cache per type. I am just wondering if this is the best place. We probably still want to cache the codecs but is this something that Hadoop does already?

Reply:
It's static, so it should be one cache. This should be equivalent to what existed before. It's caching codecs with different buffer sizes.

Reply:
Ah I see now.